File: src/main/java/net/nextlogic/airsim/gameplay/MultirotorStateUtils.java
package net.nextlogic.airsim.gameplay;

import com.google.gson.Gson;
import scala.Int;

public class MultirotorStateUtils {
    class KinematicsEstimated{
        public AirSimStructures.Vector3r position;
        public AirSimStructures.Quaternionr orientation;
    }

    class State {
        public Long timestamp;
        public Integer landed_state;
        public KinematicsEstimated kinematics_stimated;
    }

    public static AirSimStructures.Vector3r getPosition(String stateJson) {
        return parseState(stateJson).kinematics_stimated.position;
    }

    public static AirSimStructures.Quaternionr getOrientation(String stateJson) {
        return parseState(stateJson).kinematics_stimated.orientation;
    }

    public static boolean isLanded(String stateJson) {
        return parseState(stateJson).landed_state != 0;
    }

    private static State parseState(String stateJson) {
        Gson g = new Gson();
        System.out.println("Parsing state: " + stateJson);
        return g.fromJson(stateJson, State.class);
    }
}

File: src/main/kotlin/net/nextlogic/airsim/gameplay/MultirotorStateUtils.kt
//package net.nextlogic.airsim.gameplay
//
//import com.beust.klaxon.Klaxon
//
//object MultirotorStateUtils2 {
//    fun isLanded(stateJson: String): Boolean {
//        return Klaxon().parse<State>(stateJson)!!.landed_state == 0
//    }
//
//    fun getPosition(stateJson: String): AirSimStructures.Vector3r {
//        val point = Klaxon().parse<State>(stateJson)!!.kinematics_estimated.position
//
//        return AirSimStructures.Vector3r(point)
//    }
//
//    fun getOrientation(stateJson: String): AirSimStructures.Quaternionr {
//        val q = Klaxon().parse<State>(stateJson)!!.kinematics_estimated.orientation
//
//        return AirSimStructures.Quaternionr(q)
//    }
//}
//
//class Point(val x_val: Float, val y_val: Float, val z_val: Float)
//class Quaternion(val w_val: Float, val x_val: Float, val y_val: Float, val z_val: Float)
//class KinematicsEstimated(val position: Point, val orientation: Quaternion)
//class State(val timestamp: Long, val landed_state: Int, val kinematics_estimated: KinematicsEstimated)
